roles/postgresql_upgrade/tasks/main.yml

```diff
@@ -0,0 +1,62 @@
+---
+- name: Get version of postgres data directory
+  slurp:
+    path: /var/lib/pgsql/data/PG_VERSION
+  register: pg_version
+  when:
+    - ansible_os_family == "RedHat"
+    - ansible_distribution_major_version == '8'
+
+- when:
+    - ansible_os_family == "RedHat"
+    - ansible_distribution_major_version == '8'
+    - "'10' in pg_version['content']|b64decode"
+  block:
+    - name: Start postgresql
+      service:
+        name: postgresql
+        state: started
+
+    - name: Collect postgres data
+      shell: psql -lt | grep -E "^\s+postgres"
+      become: true
+      become_user: postgres
+      register: postgres_output
+
+    - name: Set collate and ctype facts
+      set_fact:
+        collate: "{{ postgres_output.stdout.split('|')[3] | trim }}"
+        ctype: "{{ postgres_output.stdout.split('|')[4] | trim }}"
+
+    - name: Stop postgresql
+      service:
+        name: postgresql
+        state: stopped
+
+    - name: Remove data directory entry from postgresql.conf
+      lineinfile:
+        path: /var/lib/pgsql/data/postgresql.conf
+        regexp: '^data_directory ='
+        line: ''
+
+    - name: Reset postgresql module
+      dnf:
+        name: '@postgresql:12'
+        state: present
+
+    - name: Update postgresql packages  # noqa package-latest
```

Comment:
This is odd. For me a `dnf module install postgresql:12` also upgraded all packages to the new version. Then it was sufficient for me to ensure `postgresql-update` was installed.